crates/paymaster-execution/src/tokens/mod.rs

```diff
@@ -0,0 +1,344 @@
+    pub fn with_base_url(base_url: &str) -> Self {
+        Self {
+            cache: SyncValue::new(CACHE_TTL),
+            client: reqwest::Client::new(),
+            base_url: base_url.to_string(),
+        }
+    }
+
+    /// Gets token info by address.
+    ///
+    /// Automatically refreshes the cache if it has expired (1-hour TTL).
+    /// Returns `None` if the token is not found.
+    pub async fn get(&self, address: Felt) -> Option<TokenInfo> {
+        let cache = self.get_cache().await.ok()?;
+        cache.get(&address).cloned()
+    }
+
+    /// Gets the token cache, refreshing if stale.
+    async fn get_cache(&self) -> Result<TokenCache, TokenServiceError> {
+        self.cache
+            .read_or_refresh({
+                let this = self.clone();
+                move || Box::pin(async move { this.fetch_all_tokens().await })
+            })
+            .await
+    }
+
+    /// Fetches all tokens from the API.
+    async fn fetch_all_tokens(&self) -> Result<TokenCache, TokenServiceError> {
+        let mut all_tokens = Vec::new();
+        let mut page = 0;
+        let page_size = 1000;
+
+        loop {
+            let url = format!("{}/v1/starknet/tokens?page={}&size={}", self.base_url, page, page_size);
+
+            let response = self
+                .client
+                .get(&url)
+                .send()
+                .await
+                .map_err(|e| TokenServiceError::HttpError(e.to_string()))?;
+
+            if !response.status().is_success() {
+                return Err(TokenServiceError::HttpError(format!("API returned status {}", response.status())));
+            }
+
+            let page_response: PageTokenDto = response
+                .json()
+                .await
+                .map_err(|e| TokenServiceError::ParseError(e.to_string()))?;
+
+            all_tokens.extend(page_response.content);
+
+            // Check if we've fetched all pages
+            if page_response.number + 1 >= page_response.total_pages {
+                break;
+            }
+
+            page += 1;
+        }
+
+        // Build the cache
+        let mut cache = HashMap::new();
+        for token in all_tokens {
+            // Parse address to Felt for consistent lookup (API always returns hex format)
+            if let Ok(felt) = Felt::from_hex(&token.address) {
+                cache.insert(felt, token);
+            } else {
+                warn!("Failed to parse token address: {}", token.address);
+            }
+        }
+
+        debug!("Token cache refreshed with {} tokens", cache.len());
+        Ok(cache)
+    }
+
+    /// Normalizes a token amount using the token's decimals.
+    ///
+    /// For example, with 18 decimals:
+    /// - 1000000000000000000 (1e18) becomes 1.0
+    /// - 100000000000000000 (1e17) becomes 0.1
+    pub async fn normalize_amount(&self, address: Felt, raw_amount: u128) -> Option<f64> {
```

Comment:
This is a utility method that should be generalized. We do that in other places as well